### PR TITLE
[g 5] Getting authentication token from Central Registry

### PR DESCRIPTION
**Pull request by tw-mahendra**

### CODE REVIEW

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        }
+    };
+
+    @Bean
+    public SecurityWebFilterChain securityWebFilterChain(
+            ServerHttpSecurity httpSecurity,
+            ReactiveAuthenticationManager authenticationManager,
+            ServerSecurityContextRepository securityContextRepository) {
+        final String[] WHITELISTED_URLS = {"/**.json",
+                "/ValueSet/**.json",
```
Comment (angshu) [inline src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: why is this required?

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                "/**.html",
+                "/**.js",
+                "/**.yaml",
+                "/**.css",
+                "/**.png",
+                "/health-information/fetch/**/attachments/**",
```
Comment (angshu) [inline src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: why is this required?

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                "/sessions",
```
Comment (angshu) [inline src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: and this?

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                "/config"};
```
Comment (angshu) [inline src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: and this?

File: build.gradle [hunk 5]
```diff
@@ -24,13 +24,15 @@ dependencies {
...
+	implementation 'org.springframework.boot:spring-boot-starter-security'
 	compileOnly 'org.projectlombok:lombok'
 	annotationProcessor 'org.projectlombok:lombok'
+	implementation 'log4j:log4j:1.2.17'
 	testImplementation('org.springframework.boot:spring-boot-starter-test') {
 		exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
 	}
 	testImplementation 'io.projectreactor:reactor-test'
-	compile 'net.logstash.logback:logstash-logback-encoder:6.3'
+	implementation 'com.nimbusds:nimbus-jose-jwt:8.15'
```
Comment (angshu) [inline build.gradle:36]: is there a reason to remove the logback library? Check with Apoorva/Ajay, they would be requiring this for logging!

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+import java.util.List;
+import java.util.Map;
+
+@Configuration
+@EnableWebFluxSecurity
+public class SecurityConfiguration {
+
+    private static final List<Map.Entry<String, HttpMethod>> SERVICE_ONLY_URLS = new ArrayList<>() {
+        {
+
```
Comment (angshu) [inline src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: remove whitespaces

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    };
+
+    @Bean
+    public SecurityWebFilterChain securityWebFilterChain(
+            ServerHttpSecurity httpSecurity,
+            ReactiveAuthenticationManager authenticationManager,
+            ServerSecurityContextRepository securityContextRepository) {
+        final String[] WHITELISTED_URLS = {"/**.json",
+                "/ValueSet/**.json",
+                "/**.html",
```
Comment (angshu) [inline src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: actually none of them are required. They are introduced because we wanted to serve the Swagger API but we can introduce them as and when required

File: src/main/resources/application.yaml [hunk 8]
```diff
@@ -1,2 +1,7 @@
 server:
   port: 8000
+clientregistry:
+  XAuthToken: c3f88ccc-2374-4a6d-bdba-63403e454979
```
Comment (angshu) [inline src/main/resources/application.yaml]: these should be all ${PROPERTY_NAME} names .. and not hardcoded. You may keep a "application-dev.yaml" file for such values.

**Review by angshu — CHANGES_REQUESTED** [review]

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java (hunk 1 above)
Comment (tw-mahendra) [reply src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: for future use we kept them as it is. however we removed now but whenever required we can add

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java (hunk 2 above)
Comment (tw-mahendra) [reply src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: as of  now we have removed these in future if required then we can add

File: build.gradle (hunk 5 above)
Comment (tw-mahendra) [reply build.gradle:36]: mistakenly removed .. added again

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java (hunk 6 above)
Comment (tw-mahendra) [reply src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: removed white spaces

File: src/main/resources/application.yaml (hunk 8 above)
Comment (tw-mahendra) [reply src/main/resources/application.yaml]: added application-dev.yaml file and removed hard coded properties

File: src/main/resources/application.yaml [hunk 9]
```diff
@@ -1,2 +1,7 @@
 server:
   port: 8000
+clientregistry:
+  XAuthToken: ${Auth_Token}
```
Comment (angshu) [inline src/main/resources/application.yaml]: Why are some vars vars in lower case?

**Review by angshu — CHANGES_REQUESTED** [review]

File: src/main/java/in/projecteka/gateway/SecurityConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            }
+            String requestPath = exchange.getRequest().getPath().toString();
+            HttpMethod requestMethod = exchange.getRequest().getMethod();
+
+            if (isCentralRegistryAuthenticatedOnlyRequest(
+                    requestPath,
+                    requestMethod)) {
+                return checkCentralRegistry(token);
+            }
+            return null; //TODO need to change
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/gateway/SecurityConfiguration.java]: You don't need this check, by default `checkCentralRegistry`

File: src/main/java/in/projecteka/gateway/clients/Caller.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+@Getter
+public class Caller {
+    private String username;
+    private Boolean isServiceAccount;
+    private String role;
+    private boolean verified;
+
+    public Optional<String> getRole() {
+        return Optional.ofNullable(role);
+    }
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/gateway/clients/Caller.java]: you don't need this.

File: src/main/java/in/projecteka/gateway/clients/model/Session.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Builder
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/gateway/clients/model/Session.java:13]: You should create separate PR to use only @Value to make classes immutable.

File: src/main/java/in/projecteka/gateway/common/CentralRegistry.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+@AllArgsConstructor
+public class CentralRegistry {
+    private final ClientRegistryClient clientRegistryClient;
+    private final ClientRegistryProperties properties;
+
+    public Mono<String> authenticate() {
+        return clientRegistryClient
+                .getTokenFor(properties.getClientId(), properties.getXAuthToken())
+                .map(session -> format("%s %s", session.getTokenType(), session.getAccessToken()));
+    }
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/gateway/common/CentralRegistry.java]: Separate PR needs to be raised for caching the token, so that we will not make a call to CR always.

File: registry.yaml [hunk 14]
```diff
@@ -1,10 +1,10 @@
...
+  - host: http://localhost:9000
     id: ncg
   - host: http://consent-manager:8000
     id: nhs
 bridges:
-  - host: http://hip:80
+  - host: http://localhost:8000
     id: 10000005
     servesAs:
       - HIP
```
Comment (ganesanarun) [inline registry.yaml]: this should not be merged.

File: src/main/java/in/projecteka/gateway/GatewayApplication.java [hunk 15]
```diff
@@ -1,17 +1,20 @@
...
 @SpringBootApplication
-@EnableConfigurationProperties({ RedisOptions.class,
-								ServiceOptions.class})
+@EnableConfigurationProperties({RedisOptions.class,
+		ServiceOptions.class,
+		ClientRegistryProperties.class})
 public class GatewayApplication {
 
 	public static void main(String[] args) {
+	    System.setProperty("spring.config.location","classpath:/registry.yaml,classpath:/application.yaml");
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/gateway/GatewayApplication.java]: Why do we need this?

File: src/main/java/in/projecteka/gateway/clients/Caller.java (hunk 11 above)
Comment (tw-mahendra) [reply src/main/java/in/projecteka/gateway/clients/Caller.java]: yes removed

File: src/main/java/in/projecteka/gateway/clients/model/Session.java (hunk 12 above)
Comment (tw-mahendra) [reply src/main/java/in/projecteka/gateway/clients/model/Session.java:13]: yes i'll do that and create separate PR for the same

File: registry.yaml (hunk 14 above)
Comment (tw-mahendra) [reply registry.yaml]: yes removed this also

File: src/main/java/in/projecteka/gateway/GatewayApplication.java (hunk 15 above)
Comment (tw-mahendra) [reply src/main/java/in/projecteka/gateway/GatewayApplication.java]: removed this..

**Review by ganesanarun — APPROVED** [review]

**Review by angshu — APPROVED** [review]
